internal/controller/nicconfigurationtemplate_controller.go

```diff
@@ -43,26 +57,262 @@ type NicConfigurationTemplateReconciler struct {
+		templates = append(templates, &template)
+	}
+
+	for _, device := range deviceList.Items {
+		node, ok := nodeMap[device.Status.Node]
+		if !ok {
+			log.Log.Info("device doesn't match any node, skipping", "device", device.Name)
+			continue
+		}
+
+		var matchingTemplates []*v1alpha1.NicConfigurationTemplate
+
+		for _, template := range templates {
+			if !deviceMatchesSelectors(&device, template, node) {
+				err := r.dropDeviceFromStatus(ctx, device.Name, template)
+				if err != nil {
+					log.Log.Error(err, "Failed to drop device name from template status", "template", template)
+				}
+
+				continue
+			}
+
+			matchingTemplates = append(matchingTemplates, template)
+		}
+
+		if len(matchingTemplates) == 0 {
+			log.Log.V(2).Info("Device doesn't match any configuration template, resetting the spec", "device", device.Name)
+			device.Spec.Configuration = nil
+			err = r.Update(ctx, &device)
+			if err != nil {
+				log.Log.Error(err, "Failed to update device's spec", "device", device)
+				return ctrl.Result{}, err
+			}
+			continue
+		} else if len(matchingTemplates) > 1 {
+			for _, template := range matchingTemplates {
+				err := r.dropDeviceFromStatus(ctx, device.Name, template)
+				if err != nil {
+					log.Log.Error(err, "Failed to drop device name from template status", "template", template)
+				}
+			}
 
-	// TODO(user): your logic here
+			templateNames := []string{}
+			for _, template := range matchingTemplates {
+				templateNames = append(templateNames, template.Name)
+			}
+			joinedTemplateNames := strings.Join(templateNames, ",")
+			err = fmt.Errorf("device matches several configuration templates: %s, %s", device.Name, joinedTemplateNames)
+			log.Log.Error(err, "Device matches several configuration templates, deleting its spec and emitting an error event", "device", device.Name)
+			return ctrl.Result{}, r.handleErrorSeveralMatchingTemplates(ctx, &device, joinedTemplateNames)
+		}
+
+		matchingTemplate := matchingTemplates[0]
+
+		if !slices.Contains(matchingTemplate.Status.NicDevices, device.Name) {
+			matchingTemplate.Status.NicDevices = append(matchingTemplate.Status.NicDevices, device.Name)
+		}
+
+		err = r.applyTemplateToDevice(ctx, &device, matchingTemplate)
+		if err != nil {
+			log.Log.Error(err, "failed to apply template to device", "template", matchingTemplate.Name, "device", device.Name)
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Try to update template's status with added / deleted devices
+	for _, template := range templates {
+		err = r.Status().Update(ctx, template)
+		if err != nil {
+			log.Log.Error(err, "failed to update template status", "template", template.Name)
+			return ctrl.Result{}, err
+		}
+	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *NicConfigurationTemplateReconciler) dropDeviceFromStatus(ctx context.Context, deviceName string, template *v1alpha1.NicConfigurationTemplate) error {
+	if slices.Contains(template.Status.NicDevices, deviceName) {
```

Comment:
nit: you can save call of Contains - Index will return -1 if the element is not found

Reply:
Thanks for the tip! Applied it